### ISSUE TITLE
sata ssd, disk.capacity = none 

### ISSUE THREAD

**Issue by Shablykinm** [truenas/py-SMART#72]
I noticed that I can't get the value of the capacity attribute.
disk = Device('/dev/sda')
print(disk.capacity)
![image](https://github.com/truenas/py-SMART/assets/39556220/3545a41c-3933-4255-8d87-d511ff774cc9)

I decided to check what data the library provides me PySMART
![image](https://github.com/truenas/py-SMART/assets/39556220/3c99ceb5-d2b0-47ba-b845-0253293e4870)

- `smartctl --scan-open`
```
C:\Windows\system32>smartctl --scan-open
/dev/sda -d ata # /dev/sda, ATA device
/dev/sdb -d sat # /dev/sdb [SAT], ATA device
- `smartctl --all \dev\sda`
C:\Windows\system32>smartctl --all /dev/sda
smartctl 7.3 2022-02-28 r5338 [x86_64-w64-mingw32-w10-b19045] (sf-7.3-1)
Copyright (C) 2002-22, Bruce Allen, Christian Franke, www.smartmontools.org

=== START OF INFORMATION SECTION ===
Model Family:     Apacer AS340 SSDs
Device Model:     Apacer AS340 240GB
Serial Number:    LKC125R001176
LU WWN Device Id: 5 02b2a2 01d1c1b1a
Add. Product Id:  mavlsata
Firmware Version: V4.25.0
User Capacity:    240 057 409 536 bytes [240 GB]
Sector Size:      512 bytes logical/physical
Rotation Rate:    Solid State Device
TRIM Command:     Available
Device is:        In smartctl database 7.3/5319
ATA Version is:   ACS-3 T13/2161-D revision 4
SATA Version is:  SATA 3.2, 6.0 Gb/s (current: 6.0 Gb/s)
Local Time is:    Fri May 26 11:16:44 2023 RTZ
SMART support is: Available - device has SMART capability.
SMART support is: Enabled

=== START OF READ SMART DATA SECTION ===
SMART overall-health self-assessment test result: PASSED

General SMART Values:
Offline data collection status:  (0x02) Offline data collection activity
                                        was completed without error.
                                        Auto Offline Data Collection: Disabled.
Self-test execution status:      (   0) The previous self-test routine completed
                                        without error or no self-test has ever
                                        been run.
Total time to complete Offline
data collection:                (    0) seconds.
Offline data collection
capabilities:                    (0x75) SMART execute Offline immediate.
                                        No Auto Offline data collection support.
                                        Abort Offline collection upon new
                                        command.
                                        No Offline surface scan supported.
                                        Self-test supported.
                                        Conveyance Self-test supported.
                                        Selective Self-test supported.
SMART capabilities:            (0x0002) Does not save SMART data before
                                        entering power-saving mode.
                                        Supports SMART auto save timer.
Error logging capability:        (0x00) Error logging NOT supported.
                                        General Purpose Logging supported.
Short self-test routine
recommended polling time:        (   2) minutes.
Extended self-test routine
recommended polling time:        (  90) minutes.
Conveyance self-test routine
recommended polling time:        (   2) minutes.

SMART Attributes Data Structure revision number: 5
Vendor Specific SMART Attributes with Thresholds:
ID# ATTRIBUTE_NAME          FLAG     VALUE WORST THRESH TYPE      UPDATED  WHEN_FAILED RAW_VALUE
  5 Reallocated_Sector_Ct   0x0033   100   100   010    Pre-fail  Always       -       0
  9 Power_On_Hours          0x0033   100   100   010    Pre-fail  Always       -       3517
 12 Power_Cycle_Count       0x0033   100   100   010    Pre-fail  Always       -       7
161 Unknown_Attribute       0x0033   100   100   010    Pre-fail  Always       -       0
164 Average_Erase_Count     0x0033   100   100   010    Pre-fail  Always       -       66135
165 Unknown_Attribute       0x0033   100   100   010    Pre-fail  Always       -       87
166 Later_Bad_Block_Count   0x0033   100   100   010    Pre-fail  Always       -       3
167 SSD_Protect_Mode        0x0033   100   100   010    Pre-fail  Always       -       64
169 Unknown_Attribute       0x0033   100   100   010    Pre-fail  Always       -       0
176 Erase_Fail_Count_Chip   0x0033   100   100   010    Pre-fail  Always       -       0
177 Wear_Leveling_Count     0x0033   100   100   010    Pre-fail  Always       -       0
178 Used_Rsvd_Blk_Cnt_Chip  0x0033   100   100   010    Pre-fail  Always       -       0
192 Unexpect_Power_Loss_Ct  0x0033   100   100   010    Pre-fail  Always       -       0
194 Temperature_Celsius     0x0033   100   100   010    Pre-fail  Always       -       47
195 Hardware_ECC_Recovered  0x0033   100   100   010    Pre-fail  Always       -       0
199 UDMA_CRC_Error_Count    0x0033   100   100   010    Pre-fail  Always       -       0
241 Total_LBAs_Written      0x0033   100   100   010    Pre-fail  Always       -       3848
242 Total_LBAs_Read         0x0033   100   100   010    Pre-fail  Always       -       1268
243 Unknown_Attribute       0x0033   100   100   010    Pre-fail  Always       -       0
244 Unknown_Attribute       0x0033   100   100   010    Pre-fail  Always       -       0
245 Unknown_Attribute       0x0033   100   100   010    Pre-fail  Always       -       0
246 Unknown_Attribute       0x0033   100   100   010    Pre-fail  Always       -       0
248 Unknown_Attribute       0x0033   100   100   010    Pre-fail  Always       -       0
249 Unknown_Attribute       0x0033   100   100   010    Pre-fail  Always       -       0
250 Read_Error_Retry_Rate   0x0033   100   100   010    Pre-fail  Always       -       0
251 Unknown_Attribute       0x0033   100   100   010    Pre-fail  Always       -       0
252 Unknown_Attribute       0x0033   100   100   010    Pre-fail  Always       -       0
253 Unknown_Attribute       0x0033   100   100   010    Pre-fail  Always       -       0
254 Unknown_SSD_Attribute   0x0033   100   100   010    Pre-fail  Always       -       0

SMART Error Log not supported

SMART Self-test log structure revision number 1
No self-tests have been logged.  [To run self-tests, use: smartctl -t]

SMART Selective self-test log data structure revision number 1
 SPAN  MIN_LBA  MAX_LBA  CURRENT_TEST_STATUS
    1        0        0  Not_testing
    2        0        0  Not_testing
    3        0        0  Not_testing
    4        0        0  Not_testing
    5        0        0  Not_testing
Selective self-test flags (0x0):
  After scanning selected spans, do NOT read-scan remainder of disk.
If Selective self-test is pending on power-up, resume after 0 minute delay.
```

**Environmental setup:**
 - Complete smartctl version smartmontools-7.3-1
 - OS Windows 10 Pro
 - Py-SMART version 1.2.4
 - PyCharm community 2023.1.2

**Comment by ralequi** [comment]
Ok, I've found 2 things for now.

1- There is an issue parsing that capacity. On my tests its readed as 240 Bytes instead of 240GB.
2- On my setup the capacity is WRONGLY readed BUT it's not None.

While I fix the first point... could you check the following values of that object?
- `_capacity`
- `_capacity_human`
- `size`
- `size_raw`

Thanks!

**Comment by Shablykinm** [comment]
![image](https://github.com/truenas/py-SMART/assets/39556220/7313f641-54e9-40f3-a1e4-725bb14cfdcc)
__
could it have something to do with the encoding or locale on the system?
I'm from Russia

**Comment by ralequi** [comment]
Yeah, you might be right... The only possibilities are "windows"-issue or "locale"-issues...

Quite strange everything is "none" or 0.

Could you try with the develop branch? ( `pip install git+https://github.com/truenas/py-SMART@develop` )
I added some special chars from your output (that spaces are not "true" spaces, for example), however probably won't be enough (since your pysmart is just not reading anything....)

If with the develop version still fails, the only thing I can think is that you debug the program. 
put a breackpoint on line device.py:1094 (on `m = re.match(r'.*:\s+([\d,. \u2019\u00a0]+)\s\D*\[?([^\]]+)?\]?', line.strip())`) and check wtf is happening there... 

copy here any raw outputs or insights you found please

**Comment by Shablykinm** [comment]
> Да, возможно, вы правы... Единственными возможными вариантами являются проблемы с "окнами" или "локалью"...
> 
> Довольно странно все "none" или 0.
> 
> Не могли бы вы попробовать с веткой разработки? ( `pip install git+https://github.com/truenas/py-SMART@develop`) Я добавил некоторые специальные символы из вашего вывода (например, пробелы не являются «настоящими» пробелами), однако, вероятно, этого будет недостаточно (поскольку ваш pysmart просто ничего не читает....)
> 
> Если с версией для разработки все еще не удается, единственное, что я могу думать, это то, что вы отлаживаете программу. Поставьте точку останова на строку device.py:1094 (на `m = re.match(r'.*:\s+([\d,. \u2019\u00a0]+)\s\D*\[?([^\]]+)?\]?', line.strip())`) и проверьте, что там происходит wtf... (и на следующих строках)
> 
> скопируйте сюда любые необработанные результаты или идеи, которые вы нашли, пожалуйста


```
PS C:\smartmqttclient> pip install git+https://github.com/truenas/py-SMART@develop
Collecting git+https://github.com/truenas/py-SMART@develop
  Cloning https://github.com/truenas/py-SMART (to revision develop) to c:\users\shablykinmy\appdata\local\temp\pip-req-build-r6f6eaib
  Running command git clone --filter=blob:none --quiet https://github.com/truenas/py-SMART 'C:\Users\shablykinmy\AppData\Local\Temp\pip-req-build-r6f6eaib'
  error: unable to create file tests/dataset/listingtests/mac_01/_-d_ata_--all_IOService__AppleACPIPlatformExpert_PCI0@0_AppleACPIPCI_RP06@1C,5_IOPP_SSD0@0_AppleAHCI_PRT0@0_IOA
HCIDevice@0_AppleAHCIDiskDriver_IOAHCIBlockStorageDevice: Filename too long
  error: unable to create file tests/dataset/singletests/sata_ssd_macos_0/_-d_ata_--all_IOService__AppleACPIPlatformExpert_PCI0@0_AppleACPIPCI_RP06@1C,5_IOPP_SSD0@0_AppleAHCI_P
RT0@0_IOAHCIDevice@0_AppleAHCIDiskDriver_IOAHCIBlockStorageDevice: Filename too long
  fatal: unable to checkout working tree
  warning: Clone succeeded, but checkout failed.
  You can inspect what was checked out with 'git status'
  and retry with 'git restore --source=HEAD :/'

  error: subprocess-exited-with-error

  × git clone --filter=blob:none --quiet https://github.com/truenas/py-SMART 'C:\Users\shablykinmy\AppData\Local\Temp\pip-req-build-r6f6eaib' did not run successfully.
  │ exit code: 128
  ╰─> See above for output.

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× git clone --filter=blob:none --quiet https://github.com/truenas/py-SMART 'C:\Users\shablykinmy\AppData\Local\Temp\pip-req-build-r6f6eaib' did not run successfully.
│ exit code: 128
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.

```
Am I doing something wrong?

**Comment by ralequi** [comment]
wow, it seems like the new pyproject.toml does not likes the pip install from git, wait a second

**Comment by Shablykinm** [comment]
> Yeah, you might be right... The only possibilities are "windows"-issue or "locale"-issues...
> 
> Quite strange everything is "none" or 0.
> 
> Could you try with the develop branch? ( `pip install git+https://github.com/truenas/py-SMART@develop` ) I added some special chars from your output (that spaces are not "true" spaces, for example), however probably won't be enough (since your pysmart is just not reading anything....)
> 
> If with the develop version still fails, the only thing I can think is that you debug the program. Put a breakpoint on line device.py:1094 (on `m = re.match(r'.*:\s+([\d,. \u2019\u00a0]+)\s\D*\[?([^\]]+)?\]?', line.strip())`) and check wtf is happening there... (and on the following lines)
> 
> copy here any raw outputs or insights you found please

while doing nothing. Set a breakpoint. I have 1091 lines
stdout?
![image](https://github.com/truenas/py-SMART/assets/39556220/d5a3f1b8-13bd-4bc7-bf1f-a52980598441)

**Comment by Shablykinm** [comment]
raw
![image](https://github.com/truenas/py-SMART/assets/39556220/28e9c92f-1f46-4c73-aafa-7e74175bd9f9)

**Comment by Shablykinm** [comment]
![image](https://github.com/truenas/py-SMART/assets/39556220/4283d6d2-1e3f-4414-b648-7b84f5fa7bcc)
I'm sorry, I completely forgot.
library gave an error on the line with utf-8 encoding.
I looked through the forums and they advised me to change to cp866
I didn't find a better solution and then I forgot I made the changes

**Comment by Shablykinm** [comment]
> pip install pySMART==1.2.5.dev23

```
C:\Users\shablykinmy\AppData\Local\Programs\Python\Python311\python.exe C:/Users/shablykinmy/AppData/Local/JetBrains/Toolbox/apps/PyCharm-C/ch-0/231.9011.38/plugins/python-ce/helpers/pydev/pydevd.py --multiprocess --qt-support=auto --client 127.0.0.1 --port 53071 --file C:\smartmqttclient\test.py 
Connected to pydev debugger (build 231.9011.38)
Traceback (most recent call last):
  File "C:\Users\shablykinmy\AppData\Local\JetBrains\Toolbox\apps\PyCharm-C\ch-0\231.9011.38\plugins\python-ce\helpers\pydev\pydevd.py", line 1496, in _exec
    pydev_imports.execfile(file, globals, locals)  # execute the script
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\shablykinmy\AppData\Local\JetBrains\Toolbox\apps\PyCharm-C\ch-0\231.9011.38\plugins\python-ce\helpers\pydev\_pydev_imps\_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "C:\smartmqttclient\test.py", line 3, in <module>
    disk = Device('/dev/sda')
           ^^^^^^^^^^^^^^^^^^
  File "C:\Users\shablykinmy\AppData\Local\Programs\Python\Python311\Lib\site-packages\pySMART\device.py", line 300, in __init__
    self.update()
  File "C:\Users\shablykinmy\AppData\Local\Programs\Python\Python311\Lib\site-packages\pySMART\device.py", line 937, in update
    raw = self.smartctl.all(
          ^^^^^^^^^^^^^^^^^^
  File "C:\Users\shablykinmy\AppData\Local\Programs\Python\Python311\Lib\site-packages\pySMART\smartctl.py", line 203, in all
    return self.generic_call(['-d', interface, '--all', disk], pass_options=True)[0]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\shablykinmy\AppData\Local\Programs\Python\Python311\Lib\site-packages\pySMART\smartctl.py", line 117, in generic_call
    return self._exec(popen_list)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\shablykinmy\AppData\Local\Programs\Python\Python311\Lib\site-packages\pySMART\smartctl.py", line 148, in _exec
    _stdout, _stderr = [i.decode('utf8') for i in proc.communicate()]
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\shablykinmy\AppData\Local\Programs\Python\Python311\Lib\site-packages\pySMART\smartctl.py", line 148, in <listcomp>
    _stdout, _stderr = [i.decode('utf8') for i in proc.communicate()]
                        ^^^^^^^^^^^^^^^^
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xa0 in position 413: invalid start byte

Process finished with exit code 1

```

**Comment by ralequi** [comment]
parsing errors might came from that change...

however, if your output it's not utf-8... anything can happens...

- try changing the encoding again (at least for a test)
- please, DUMP the call into a file of smartctl ( `smartctl -d ata --all /dev/sda  > _-d_ata_--all__dev_sda` ) and upload here as a file... Let's see if that coding is kept on the file so I can do better tests with this

**Comment by Shablykinm** [comment]
> ошибки синтаксического анализа могут возникнуть из-за этого изменения...
> 
> однако, если ваш вывод не utf-8... может случиться что угодно...
> 
> * попробуйте еще раз изменить кодировку (хотя бы для теста)
>   
>   * И скажите мне, если работает!
> * пожалуйста, скопируйте вызов в файл smartctl( `smartctl -d ata --all /dev/sda  > _-d_ata_--all__dev_sda`) и загрузите сюда в виде файла... Давайте посмотрим, сохраняется ли этот код в файле, чтобы я мог лучше протестировать его.

```
        encoding = "cp866"
        _stdout, _stderr = [i.decode(encoding) for i in proc.communicate()]
```
![image](https://github.com/truenas/py-SMART/assets/39556220/7f6454ca-0584-4b13-83b5-8a2bec228c8d)

smartctl -d ata --all /dev/sda  > _-d_ata_--all__dev_sda
the file is not uploaded to github, so uploaded it to the drive
https://drive.google.com/file/d/1FfnU8--CuJPeZ2phQSH9ZxXwc6eVDNvE/view?usp=sharing

**Comment by ralequi** [comment]
confirmed it is not utf-8. My linux says it is `ISO-8859`

Ok, give me some time to check the best way to handle this

**Comment by Shablykinm** [comment]
Went back to 1.2.4
 device.py in 1096 re:
.*:\s+([\d,.]+)\s\D*\[?([^\]]+)?\]?
most likely in my case the problem is "\d"
![image](https://github.com/truenas/py-SMART/assets/39556220/011da366-56cb-40ad-a814-d4113acbb6d1)

because I get a line
User Capacity:    240а057а409а536 bytes [240 GB]
![image](https://github.com/truenas/py-SMART/assets/39556220/49a5d63c-31d2-4af3-a080-d6887cf0800c)

I tried to make a replay, but then I realized that these are unicode characters
I did the following:
```
            if any_in(line, 'User Capacity', 'Total NVM Capacity', 'Namespace 1 Size/Capacity'):
                # TODO: support for multiple NVMe namespaces
                line2 = line.encode('ascii', 'ignore').decode()
                m = re.match(
                    r'.*:\s+([\d,.]+)\s\D*\[?([^\]]+)?\]?', line.strip().replace('а', ''))

                if m is not None:
                    tmp = m.groups()
                    self._capacity = int(
                        tmp[0].strip().replace(',', '').replace('.', ''))

                    if len(tmp) == 2 and tmp[1] is not None:
                        self._capacity_human = tmp[1].strip().replace(',', '.')
```
now i get it
raw:
'User Capacity:    240057409536 bytes [240 GB]
'
![image](https://github.com/truenas/py-SMART/assets/39556220/d0f4b35f-491a-4e13-96e3-1485dd1a4156)

**Comment by Shablykinm** [comment]
I hope this should help

**Comment by ralequi** [comment]
I don't think is that easy, I'll continue with this on Monday

**Comment by ralequi** [comment]
Hi @Shablykinm 

Sorry for the delay.
Could you try `pip install pySMART==1.2.5.dev24` ? 
I've added an automatically charset detector. It seems to work great on my files, but I would like to hear from you on the "real field".

Tell me if works so I can release 1.2.5 soon ;-)

**Comment by Shablykinm** [comment]
Ooo! Great!
ssd and hdd work fine.
It seems that now the same thing with nvme
![image](https://github.com/truenas/py-SMART/assets/39556220/ad4204c2-b288-430a-a7bc-a833964d78ae)
![image](https://github.com/truenas/py-SMART/assets/39556220/00dc999b-498c-43e0-9d3a-c1a9f7576421)

**Comment by ralequi** [comment]
I think I should replace all those problematic chars at once instead of fix each regex...

Could you please upload (at google/ondrive/whatever) that file too?
I mean the output of `smartctl -d nvme --all /dev/sda > _-d_nvme_--all__dev_sda`

On your shared picture... I'm not sure if they are exactly the same chars I've found earlier (i mean u00a0), but looks like a new strange char

Thanks!

**Comment by Shablykinm** [comment]
try)
https://drive.google.com/file/d/1z30RJwr9vj5ACnTdjefMEwwTnZcj8o1X/view?usp=sharing
Why not use decode()?
line.encode('ascii', 'ignore').decode()

**Comment by ralequi** [comment]
Yeah, your idea works!

**Comment by ralequi** [comment]
Try with `pip install pySMART==1.2.5.dev28` 

Now, finally, it must be fixed! (please? haha)

Waiting for your tests!

**Comment by Shablykinm** [comment]
maybe it's because there are no numbers before the attribute?
![image](https://github.com/truenas/py-SMART/assets/39556220/95ec83c9-141e-44a0-aa23-97bfb71aa8ed)

**Comment by Shablykinm** [comment]
![image](https://github.com/truenas/py-SMART/assets/39556220/df0cd773-4b10-441b-9a22-98f527a7b207)

**Comment by ralequi** [comment]
Sorry, I confess py-smart API is very confusing, it requires a rework (that will come soon btw)

attributes field is a legacy (& soon deprecated) for **ATA disks ONLY**.

check on field `if_attributes` for nvme devices.
You should find something like:

```json
        "if_attributes": {
            "critialWarning": null,
            "temperature": 38,
            "availableSpare": 100,
            "availableSpareThreshold": 10,
            "percentageUsed": 1,
            "dataUnitsRead": 21600976,
            "bytesRead": 11000000000000,
            "dataUnitsWritten": 16184623,
            "bytesWritten": 8279999999999,
            "hostReadCommands": 272980851,
            "hostWriteCommands": 210904247,
            "controllerBusyTime": 5077,
            "powerCycles": 908,
            "powerOnHours": 3387,
            "unsafeShutdowns": 46,
            "integrityErrors": 0,
            "errorEntries": 0,
            "warningTemperatureTime": null,
            "criticalTemperatureTime": 0,
            "errors": [],
            "tests": [],
            "criticalWarning": 0
        },
```

**Comment by Shablykinm** [comment]
![image](https://github.com/truenas/py-SMART/assets/39556220/b34b606c-c91c-4ab0-83be-0d5b776af1f1)
thankyou

**Comment by ralequi** [comment]
Glad to help!

Thank you for take your time reporting this issues!

upcoming 1.2.5 soon.

I'm closing this issue, please, reopen if you find something new! (or a new issue)
